lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,257 @@
+	var err error
+	for _, txn := range txnData.Txns {
+		if fee, err = SafeUint64().Add(fee, txn.TxnFeeNanos); err != nil {
+			return 0, errors.Wrapf(err, "AtomicTxnsMetadata.GetTotalFee: Problem computing fee")
+		}
+	}
+	return fee, nil
+}
+
+func (txnData *AtomicTxnsMetadata) GetTxnType() TxnType {
+	return TxnTypeAtomicTxns
+}
+
+func (txnData *AtomicTxnsMetadata) ToBytes(preSignature bool) ([]byte, error) {
+	var data []byte
+	data = append(data, UintToBuf(uint64(len(txnData.Txns)))...)
+	for _, txn := range txnData.Txns {
+		txnBytes, err := txn.ToBytes(preSignature)
+		if err != nil {
+			return nil, errors.Wrapf(err, "AtomicTxnsMetadata.ToBytes: Problem serializing txn: ")
+		}
+		data = append(data, UintToBuf(uint64(len(txnBytes)))...)
+		data = append(data, txnBytes...)
+	}
+	return data, nil
+}
+
+func (txnData *AtomicTxnsMetadata) FromBytes(data []byte) error {
+	rr := bytes.NewReader(data)
+
+	// NumTxns
+	numTxns, err := ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrap(err, "AtomicTxnsMetadata.FromBytes: Problem reading NumTxns")
+	}
+	txnData.Txns, err = SafeMakeSliceWithLength[*MsgDeSoTxn](numTxns)
+	if err != nil {
+		return errors.Wrap(err, "AtomicTxnsMetadata.FromBytes: Problem allocating Txns")
+	}
+	for ii := uint64(0); ii < numTxns; ii++ {
+		txnData.Txns[ii] = &MsgDeSoTxn{}
+		numTxnBytes, err := ReadUvarint(rr)
+		if err != nil {
+			return errors.Wrap(err, "AtomicTxnsMetadata.FromBytes: Problem reading NumTxnBytes")
+		}
+		txnBytes, err := SafeMakeSliceWithLength[byte](numTxnBytes)
+		if err != nil {
+			return errors.Wrap(err, "AtomicTxnsMetadata.FromBytes: Problem allocating TxnBytes")
+		}
+		if _, err = io.ReadFull(rr, txnBytes); err != nil {
+			return errors.Wrap(err, "AtomicTxnsMetadata.FromBytes: Problem reading TxnBytes")
+		}
+		if err = txnData.Txns[ii].FromBytes(txnBytes); err != nil {
+			return errors.Wrapf(err, "AtomicTxnsMetadata.FromBytes: Problem parsing txn %d: ", ii)
+		}
+	}
+	return nil
+}
+
+func (txnData *AtomicTxnsMetadata) New() DeSoTxnMetadata {
+	return &AtomicTxnsMetadata{}
+}
+
+func (bav *UtxoView) _connectAtomicTxns(
+	txn *MsgDeSoTxn, txHash *BlockHash, blockHeight uint32, blockTimestampNanoSecs int64, verifySignatures bool,
+	ignoreUtxos bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error,
+) {
+
+	// Verify the transaction is well-formed.
+	// TODO: verify
+
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight {
+		return 0, 0, nil, errors.Wrap(RuleErrorAtomicTxnBeforeBlockHeight, "_connectAtomicTxns")
+	}
+
+	if txn.TxnMeta.GetTxnType() != TxnTypeAtomicTxns {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectAtomicTxns: TxnMeta type: %v", txn.TxnMeta.GetTxnType().GetTxnString())
+	}
+
+	var totalInput, totalOutput, totalFees uint64
+	var utxoOps []*UtxoOperation
+	// Create copy of view. We apply all transactions to the copy first, and then
+	// if successful, apply to the real view.
+	viewCopy, err := bav.CopyUtxoView()
```

Comment:
I don't think you need to do this viewCopy thing. Why not just error if you fail on one of the txns? It's already understood that the view will be compromised if an error happens in the middle of connecting a txn right?